README.md
# CHANGE ME 

## License

The gem is available as open source under the terms of the [Apache-2.0 License](http://opensource.org/licenses/Apache-2.0).



CHANGELOG.md
# Change log

This document represents a high-level overview of changes made to this project.
It will not list every miniscule change, but will allow you to view - at a
glance - what to expact from upgrading to a new version.

## [unpublished]

### Added

### Changed

### Fixed

### Security

### Deprecated

### Removed


## [0.4.5] - 2018-09-09

### Fixed

- Handle ID 64s exceeding allowed range, by adding/subtracting multiples of
  range until fitting.


## [0.4.4] - 2018-03-25

### Fixed

- Handle vanity URLs which can't be part of a valid URI by throwing an
  exception instead of failing when creating an URL.


## [0.4.3] - 2017-11-30

### Fixed

- Handle non-ASCII vanity URLs by throwing an exception instead of failing when
  creating an URL.


## [0.4.2] - 2017-11-25

### Fixed

- Fix `SteamID::Parser#from_string` throwing an exception when resolving a
  vanity URL.


## [0.4.1] - 2017-11-19

### Fixed

- Missing functionality of `SteamID.from_string`.


## [0.4.0] - 2017-11-18

### Added

- `SteamID::SteamID` class representing a single Steam ID.
- `Parser` class allowing to parse various formats of Steam IDs into their
  respective account ID.
- `SteamID#from_string` convenience method.


## [0.3.0] - 2017-05-30

### Added

- Support for community URLs containing Steam ID3.


## [0.2.0] - 2017-03-07

### Added

- Support for brackets in SteamID 3 format, e.g. `[U:1:123456]`


## [0.1.2] - 2017-01-15

### Added

### Changed

### Fixed

- Fixed handling of account IDs starting with 765.


## [0.1.1] - 2017-01-02

### Fixed

- Implements missing SteamID::SteamID.from_string method.


## [0.1.0] - 2017-01-02

### Added

- Module to allow conversion of SteamIDs into account IDs suitable for e.g.
  Hive 2 API calls.

  Supported are:
  - Steam ID
  - Steam ID 3
  - Steam ID 64
  - Community URL
  - Profile URL


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'steam_id'


spec/steam-id/module_spec.rb
# coding: utf-8

RSpec.describe SteamID do
  describe '#from_string' do
    it 'should resolve the ID into a string' do
      expect(SteamID.from_string('STEAM_0:0:24110655').account_id).to eq 48221310
    end
  end
end


spec/steam-id/parser_spec.rb
# coding: utf-8

module SteamID
  RSpec.describe Parser do
    # API calls are mocked out, hence we can use a fake key.
    let(:parser) { Parser.new(api_key: 'EB9E950DED9AEADA23174D145A019ECD') }

    STEAM_ID         = 'STEAM_0:0:24110655'
    STEAM_ID_MIXED   = 'SteAM_0:0:24110655'
    STEAM_ID_SHORT   = 'STEAM_1:0:2691362'

    STEAM_ID_3          = 'U:1:48221310'
    STEAM_ID_3_LOWER    = 'u:1:48221310'
    STEAM_ID_3_SHORT    = 'U:1:4584616'
    STEAM_ID_3_BRACKETS = '[U:1:48221310]'

    STEAM_ID_64      = '76561198008487038'

    COMMUNITY_URL = 'http://steamcommunity.com/profiles/76561198008487038'
    COMMUNITY_URL_STEAM_ID_3 = 'http://steamcommunity.com/profiles/[U:1:48221310]'

    describe '::from_string' do
      # Detailed specs per-type further down.
      it 'supports Steam ID as input' do
        steam_id = parser.from_string(STEAM_ID)
        expect(steam_id.account_id).to eq 48221310
      end

      it 'supports Steam ID 3 as input' do
        steam_id = parser.from_string(STEAM_ID_3)
        expect(steam_id.account_id).to eq 48221310
      end

      it 'supports SteamID 64 as input' do
        steam_id = parser.from_string(STEAM_ID_64)
        expect(steam_id.account_id).to eq 48221310
      end

      it 'supports community URLs as input' do
        steam_id = parser.from_string(COMMUNITY_URL)
        expect(steam_id.account_id).to eq 48221310
      end

      it 'supports vanity URLs as input' do
        expect(::SteamId).to receive(:resolve_vanity_url) { 10000 }

        steam_id = parser.from_string('random-string')
        expect(steam_id.account_id).to eq 10000
      end
    end

    describe '::from_steam_id' do
      it 'supports SteamID as input' do
        steam_id = parser.from_steam_id(STEAM_ID)
        expect(steam_id.account_id).to eq 48221310

        steam_id = parser.from_steam_id(STEAM_ID_SHORT)
        expect(steam_id.account_id).to eq 5382724
      end

      it 'supports SteamID 3 as input' do
        steam_id = parser.from_steam_id(STEAM_ID_3)
        expect(steam_id.account_id).to eq 48221310

        steam_id = parser.from_steam_id(STEAM_ID_3_BRACKETS)
        expect(steam_id.account_id).to eq 48221310

        steam_id = parser.from_steam_id(STEAM_ID_3_SHORT)
        expect(steam_id.account_id).to eq 4584616
      end

      it 'is not case-sensitive' do
        steam_id = parser.from_steam_id(STEAM_ID_MIXED)
        expect(steam_id.account_id).to eq 48221310

        steam_id = parser.from_steam_id(STEAM_ID_3_LOWER)
        expect(steam_id.account_id).to eq 48221310
      end

      context 'with a Steam ID 64' do
        it 'supports those in the normal range as input' do
          steam_id = parser.from_steam_id(STEAM_ID_64)
          expect(steam_id.account_id).to eq 48221310
        end

        it 'supports those lowered by multiples of 2**32' do
          id = STEAM_ID_64.to_i

          steam_id = parser.from_steam_id(id - 1 * 2**32)
          expect(steam_id.account_id).to eq 48221310

          steam_id = parser.from_steam_id(id - 3 * 2**32)
          expect(steam_id.account_id).to eq 48221310
        end

        it 'supports those increased by multiples of 2**32' do
          id = STEAM_ID_64.to_i

          steam_id = parser.from_steam_id(id + 1 * 2**32)
          expect(steam_id.account_id).to eq 48221310

          steam_id = parser.from_steam_id(id + 4 * 2**32)
          expect(steam_id.account_id).to eq 48221310
        end
      end

      it 'supports an account ID as input' do
        steam_id = parser.from_steam_id(48221310)
        expect(steam_id.account_id).to eq 48221310

        # Special case, starts with 765, like a Steam ID
        steam_id = parser.from_steam_id(7654321)
        expect(steam_id.account_id).to eq 7654321
      end

      it 'raises an exception upon invalid input' do
        expect { parser.from_steam_id('foobar') }.to raise_error(ArgumentError)
      end
    end

    describe '::from_community_url' do
      it 'supports community URLs as input' do
        steam_id = parser.from_community_url(COMMUNITY_URL)
        expect(steam_id.account_id).to eq 48221310
      end

      it 'supports community URLs with parser3 as input' do
        steam_id = parser.from_community_url(COMMUNITY_URL_STEAM_ID_3)
        expect(steam_id.account_id).to eq 48221310
      end

      it 'raises an exception upon invalid input' do
        expect { parser.from_community_url('http://google.com/profiles/foo') }.to raise_error(ArgumentError)
      end
    end

    describe '::from_vanity_url' do
      it 'supports custom URLs as input' do
          expect(::SteamId).to receive(:resolve_vanity_url) { 54321 }

          steam_id = parser.from_vanity_url('some-test')
          expect(steam_id.account_id).to eq 54321
      end

      it 'supports full custom URLs as input' do
          expect(::SteamId).to receive(:resolve_vanity_url) { 54321 }
          expect(parser).to receive(:from_steam_id) { 12345 }

          expect(parser.from_vanity_url('http://steamcommunity.com/id/some-test')).to eq 12345
      end

      it 'raises an exception if it could not resolve the ID' do
          expect(::SteamId).to receive(:resolve_vanity_url) { nil }

          expect{ parser.from_vanity_url('this-better-not-be-a-valid-url') }.to raise_error(ArgumentError)
      end

      it 'raises an exception on non-ascii inputs' do
        expect { parser.from_vanity_url('CЯaZyCAT') }.to raise_error ArgumentError
      end

      it 'raises an exception on invalid inputs' do
        expect { parser.from_vanity_url('#1 best') }.to raise_error ArgumentError
        expect { parser.from_vanity_url('https://steamcommunity.com/id/#1 best') }.to raise_error ArgumentError
      end
    end
  end
end


lib/steam-id/parser.rb
# coding: utf-8

require 'steam-condenser'

module SteamID
  # Create SteamID objects based on numeric or string-based Steam ID inputs.
  # @see SteamID
  class Parser
    STEAM_CUSTOM_URL_BASE = 'https://steamcommunity.com/id/%s'

    # Pattern to match classical Steam IDs (STEAM_0:...)
    PATTERN_STEAM_ID = /^STEAM_[0-9]:([0-9]):([0-9]+)$/i
    # Pattern to match Steam ID 3 format ([U:1:...])
    PATTERN_STEAM_ID_3 = /^\[?U:([0-9]{1,2}):([0-9]+)\]?$/i
    # Pattern to match Steam ID 64 aka profile ID format (765...):
    # Must have at least 14 digits, else ID_64 - OFFSET would be < 0.
    PATTERN_STEAM_ID_64 = /^(765[0-9]{11,})$/
    # Pattern to match plain account IDs.
    # Not sure what its valid range is - exists with at least 7 and 8 numbers.
    PATTERN_ACCOUNT_ID = /^[0-9]+$/

    # Pattern to match ID64-based profile URLs.
    PATTERN_COMMUNITY_URL = /^https?:\/\/steamcommunity\.com\/profiles\/(765[0-9]{11,})\/?$/
    # Pattern to match ID3-based profile URLs.
    PATTERN_COMMUNITY_URL_STEAM_ID_3 = /^https?:\/\/steamcommunity\.com\/profiles\/\[U:([0-9]{1,2}):([0-9]+)\]$/i
    # Pattern to match custom-URL-based profile URLs.
    PATTERN_CUSTOM_URL = /^https?:\/\/steamcommunity\.com\/id\/([^\/]+)\/?$/

    # Initialize parser
    # @param api_key [String] Key for Steam web API. `nil` to disable API
    #   functionality.
    def initialize(api_key: nil)
      @api_key = api_key
      if api_key
        ::WebApi.api_key = @api_key.to_s
      end
    end

    # Create SteamID object based on any kind of recognized Steam ID.
    # @param s [String]: Any form of known Steam ID, profile URL, or 'vanity'/
    #   custom URL.
    # @see Parser#from_steam_id Supported formats of Steam IDs
    # @see Parser#from_community_url Supported formats of community URLs
    # @see Parser#from_vanity_url Supported formats of vanity URLs
    # @return [SteamID]
    # @raise [ArgumentError] If the supplied string was not a valid custom URL.
    # @raise [WebApiError] If the Steam API returned an error.
    def from_string(s)
      account_id = nil

      # Todo: Refactor
      begin
        # Checking for Steam ID first. Most restrictive check, and also does
        # not require a call to Steam Web API.
        account_id = from_steam_id(s)
      rescue ArgumentError
        begin
          # Community URL afterwards, does not require an API call either.
          account_id = from_community_url(s)
        rescue ArgumentError
          begin
            # Trying to resolve as custom/vanity URL now.
            account_id = from_vanity_url(s)
          rescue ArgumentError
          end
        end
      end

      if account_id.nil?
        raise ArgumentError, "Could not convert #{ s } to account id."
      else
        account_id
      end
    end

    # Create SteamID object based on Steam ID.
    # @param id [String] Steam ID, either of:
    #   - Steam ID: STEAM_0:0:24110655
    #   - Steam ID 3: U:1:48221310
    #   - Steam ID 64: 76561198008487038
    #   - Account ID: 48221310
    # @return [SteamID]
    # @raise [ArgumentError] If the supplied string could not be converted to
    #   an account ID.
    def from_steam_id(id)
      # In case we get a fixnum.
      id = id.to_s

      # https://developer.valvesoftware.com/wiki/SteamID#Format
      PATTERN_STEAM_ID.match(id) do |m|
        return SteamID.new(m[1].to_i + m[2].to_i * 2)
      end

      PATTERN_STEAM_ID_3.match(id) do |m|
        return SteamID.new(m[2].to_i)
      end

      PATTERN_STEAM_ID_64.match(id) do |m|
        id = steam_id_64_to_account_id(m[1].to_i)
        return SteamID.new(id)
      end

      # Matching this one last, as not to catch an ID 64 on accident.
      PATTERN_ACCOUNT_ID.match(id) do |m|
        return SteamID.new(id.to_i)
      end

      # If we get until here, we did not match any regex.
      raise ArgumentError, "#{ id.inspect } is not a supported SteamID."
    end

    # Create SteamID object based on community URL.
    # @param url [String] community URL: http://steamcommunity.com/profiles/76561198008487038
    # @return [SteamID]
    # @raise [ArgumentError] If the supplied string was not a valid community URL.
    def from_community_url(url)
      PATTERN_COMMUNITY_URL.match(url) do |m|
        return self.from_steam_id(m[1])
      end

      PATTERN_COMMUNITY_URL_STEAM_ID_3.match(url) do |m|
        return self.from_steam_id(m[2])
      end

      raise ArgumentError, "#{ url.inspect } is not a supported community URL."
    end

    # Create SteamID object based on custom URL. Note that this requires the
    # API key to be set.
    # @param url [String] custom URL - must be ASCII: 
    #   - http://steamcommunity.com/id/some-custom-url
    #   - some-custom-url
    # @return [SteamID]
    # @raise [ArgumentError] If the supplied string was not a valid custom URL.
    # @raise [WebApiError] If the Steam API returned an error.
    def from_vanity_url(url)
      PATTERN_CUSTOM_URL.match(url) do |m|
        url = m[1]
      end

      begin
        URI(STEAM_CUSTOM_URL_BASE % url)
      rescue URI::InvalidURIError
        raise ArgumentError, "#{ url } can't be part of a valid URI."
      end

      steam_id = SteamId.resolve_vanity_url(url)
      if steam_id.nil?
        raise ArgumentError, "#{ url } was not a valid custom URL."
      else
        from_steam_id(steam_id)
      end
    end

    private
    def steam_id_64_to_account_id(id_64)
      account_id = id_64 - SteamID::ID_64_OFFSET

      # Assuming the range to be [1, 2**32]. But in reality, who knows.
      # /profiles/#{ ID_64_OFFSET } URL (as well as offsets of 2**32 thereof)
      # currently redirect to one's signed-in profile, so seems to have some
      # meta meaning.
      while account_id < 1
        account_id += 2**32
      end

      while account_id > 2**32
        account_id -= 2**32
      end

      account_id
    end
  end
end
